Allow perfcollect to collect data for fix amount of time

shhsu opened the pull request:
User would not be required to manually press ctrl+c to terminate collection.

Similar to
perf record sleep 10

The user can do
perfcollect collect profile.data -collectsec 10

brianrob commented:
Peter, thanks for making these changes.

I did a quick test with and without the collectsec parameter, and both of them seem to have issues.  With the collectsec parameter, it appears that collection gets stopped, but no zip file is generated.  The files are still sitting in the /tmp directory.

Without the collectsec parameter, everything in the current working directory gets zipped up into the zip file and deleted out of the directory.

Can you look into these issues?  Thanks.

shhsu commented:
Brian, what's the command you are using?
I have been using the new script with or without -collectsec on my end until now. Maybe we are using different list of parameters and that might have been the issue.

shhsu commented:
Also, what's your linux tool's version?

shhsu commented:
actually I was reproing the issue now. Investigating...

shhsu commented:
Apologies for the oversights.
* The -collectsec case problem is due to a typo, fixed
* The non-collectsec case is due to a odd situation where sometimes the code continues even though ctrl+c is called and CTRLC_Handler being called two times. I am unable to repro it since I rebooted my linux machine. I added an if statement back to fix the issue but I am unsure if I need to also worry about thread safety.

So treat this as a WIP update to this PR for now

shhsu commented:
Apologies for the oversight

* -perfcollect case was broken because of a typo I introduced when fixing PR issue. Fixed
* non-perfcollect case was broke due to the wrong assumption. When ctrl+c is trapped the code actually continues executing.

Fixed both issues and tested the following 3 scenarios:

sudo ./perfcollect collect perf (stopped with ctrl-c)
sudo ./perfcollect collect perf -collectsec 1
sudo ./perfcollect collect perf -collectsec 10 (stopped early with ctrl-c)

shhsu commented:
Apologies for the oversight

* -perfcollect case was broken because of a typo I introduced when fixing PR issue. Fixed
* non-perfcollect case was broke due to the wrong assumption. When ctrl+c is trapped the code actually continues executing.

Fixed both issues and tested the following 3 scenarios:

sudo ./perfcollect collect perf (stopped with ctrl-c)
sudo ./perfcollect collect perf -collectsec 1
sudo ./perfcollect collect perf -collectsec 10 (stopped early with ctrl-c)

shhsu commented:
There doesn't seem to be a scenario where usePerf != 1. I am not sure how this change affect the scenario in line 1459. (which seem to be dead code) Please verify.

brianrob commented:
Yes, you’re right – by default usePerf = 1 in all cases.  I’ve left this in mostly for testing and debugging at this point.

From: Peter Hsu [mailto:notifications@github.com]
Sent: Friday, February 26, 2016 1:59 PM
To: dotnet/corefx-tools <corefx-tools@noreply.github.com>
Cc: Brian Robbins <brianrob@microsoft.com>
Subject: Re: [corefx-tools] Allow perfcollect to collect data for fix amount of time (#22)


There doesn't seem to be a scenario where usePerf != 1. I am not sure how this change affect the scenario in line 1459. (which seem to be dead code) Please verify.

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/corefx-tools/pull/22#issuecomment-189500910>.

brianrob commented:
Peter, thanks for implementing this.  Looks good to me.
